### ISSUE TITLE
Limit the frequency of context providers being updated.

### ISSUE THREAD

**Issue by matthusby** [TabbyML/tabby#3016]
I have a few busy git repos so they end up getting pulled and indexed all day.  Each on them take like 15 or 20 minutes to get re-indexed so my server ends up spending all day running at full load.  I did switch the embedding to use the GPU but its still not fast enough to keep up.

I don't think I get much help from have context that was updated in the last hour, so I would like to be able to just fetch new code and process it once a day (Or some interval that I can set?)

---
Please reply with a 👍 if you want this feature.

**Comment by wsxiaoys** [comment]
Thank you for reporting the issue. Could you provide more details about the size of your repository, such as the total number of files and the number of files changed per hour?

Tabby builds its index incrementally, meaning that embeddings should only be computed for documents that have changed. Therefore, a 10-20 minute reindexing time suggests that the changes are quite substantial

**Comment by matthusby** [comment]
> Tabby builds its index incrementally

Hmm, I have some problem here then as its doing a full index every time it runs.  The changes are just normal stuff, like from a couple lines to a few hundred and a couple files.

**Comment by wsxiaoys** [comment]
Could you please provide more details about your setup? For example, the version number, and how Tabby is deployed, among other relevant information

**Comment by comxd** [comment]
I also find that is very long to execute, whether for Git repositories or for "web" crawls. For GIT repositories, the crawl is triggered by a hook at the time of a commit?

I use tabby 0.16.1, in a docker compose stack:
```
    tabby:
        restart: always
        image: tabbyml/tabby
        command: serve --model DeepseekCoder-6.7B --chat-model Qwen2-1.5B-Instruct --device cuda
        volumes:
            - "$HOME/.tabby:/data"
            - ".:/project-repository"
        ports:
            - "8099:8080"
        deploy:
            resources:
                reservations:
                    devices:
                        - driver: nvidia
                          count: 1
                          capabilities: [gpu]
```

System:
```
(K)Ubuntu 24.04
16 × 11th Gen Intel® Core™ i9-11900K @ 3.50GHz
62,4 GiB of RAM
NVIDIA RTX A4000/PCIe/SSE2
ThinkStation P350
Kernel version 6.8.0-41-generic (64-bit)
X11
---
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Jan__6_16:45:21_PST_2023
Cuda compilation tools, release 12.0, V12.0.140
Build cuda_12.0.r12.0/compiler.32267302_0
---
VGA compatible controller [0300]: NVIDIA Corporation GA104GL [RTX A4000] [10de:24b0] (rev a1)
        Subsystem: NVIDIA Corporation GA104GL [RTX A4000] [10de:14ad]
```

Some outputs:

Git, https://github.com/symfony/symfony, Duration: 57 minutes:
```
2024-08-28T22:41:27.112691+00:00 [INFO]: Building source code index: https://github.com/symfony/symfony
...
2024-08-28T23:39:05.143613+00:00 [INFO]: Job completed successfully
```

Web, https://ymfony.com/doc/current/, Duration: 20 minutes:
```
2024-08-29T00:12:58.331083+00:00 [INFO]: Starting doc index pipeline for https://symfony.com/doc/current/
...
2024-08-29T00:33:28.798509+00:00 [INFO]: Crawled 866 documents from 'https://symfony.com/doc/current/'
2024-08-29T00:33:29.038243+00:00 [INFO]: Job completed successfully
```

Web, https://ckeditor.com/docs/ckeditor5/latest/, Duration: 2 hours:
```
2024-08-29T15:39:38.796792+00:00 [INFO]: Starting doc index pipeline for https://ckeditor.com/docs/ckeditor5/latest/
...
2024-08-29T17:39:38.803056+00:00 [WARN]: Crawled for url: https://ckeditor.com/docs/ckeditor5/latest/ timeout after 7200 seconds
2024-08-29T17:39:38.803094+00:00 [INFO]: Job completed successfully
```

Currently indexing from a long time with gith CPU usage:
```
# date
jeu. 29 août 2024 21:43:47 CEST

# ps aux | grep katana
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      142481 49.7  0.1 3021080 129232 ?      Sl   17:39 121:27 katana -u https://ckeditor.com/docs/ckeditor5/latest/ -jsonl -mdc starts_with(endpoint, "https://ckeditor.com/docs/ckeditor5/latest/") -depth 9999
root      167876  932  1.6 6234976 1098488 ?     Rl   19:39 1157:47 katana -u https://docs.docker.com/reference/ -jsonl -mdc starts_with(endpoint, "https://docs.docker.com/reference/") -depth 9999
root      198770  403  0.1 2804076 97156 ?       Sl   21:39  16:49 katana -u https://docs.docker.com/manuals/ -jsonl -mdc starts_with(endpoint, "https://docs.docker.com/manuals/") -depth 9999

```

**Comment by matthusby** [comment]
Version:
`Version 0.17.0-dev.0`

Deployment:
`tabby serve` in a screen session

So I did a 2 things and now it seems to be working as intended - as in only indexing the changes.
I just deleted everything in my `.tabby` folder and started over, like re adding my GH creds and adding the repos back in.
I compiled for release instead of just using the debug build.

I will continue poking at it as see if I can replicate the problem I was having.

**Comment by wsxiaoys** [comment]
@comxd - The first execution is expected to be slow because we need to compute tokens/embeddings for all the code. I would imagine a second run on the git repo should be much faster.

The web crawler is a bit different; a re-run will always recrawl all documents to ensure freshness.

---

@matthusby Thanks! Let me know if you encounter any further issues!

**Comment by comxd** [comment]
Thanks for your reply @wsxiaoys.
I understand, but a lot of job (web context provider) was failed (timeout 2H) but the job reported as success.
Also, no more information was provided in the stderr (this tab is always empty).
I think is a problem here, ask me if you want more details.

Regardings my suggestions for the web context provider is to allow is to configure some Katana parameters inside the UI:
* -timeout int
* -depth int
* -exclude string[]
* -headers string[]
* -ignore-query-params
* all FILTER and RATE-LIMIT options

Also to simplify this implementation, I think a simple text field allowing to set all parameters we want can be useful.

**Comment by matthusby** [comment]
@wsxiaoys everything has been running perfectly for a couple days now, so whatever state I found myself in before was fixed by blowing away my config and starting over (or switching to a release build).  I had done a lot of experimenting with config options and switching embedding models etc. 

I am going to close this issue, but if I do find it happening again I will get in touch to see about debugging it.
